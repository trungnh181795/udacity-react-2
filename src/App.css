body {
  background-color: #141416;
}