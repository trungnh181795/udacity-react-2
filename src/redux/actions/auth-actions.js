import { LOGOUT_AUTHED_USER, SET_AUTHED_USER } from "../constants/auth-constants";

export const setAuthedUser = (authedUser) => {
    return {
        type: SET_AUTHED_USER,
        authedUser,
    };
}

export const logoutAuthedUser = () => {
    return {
        type: LOGOUT_AUTHED_USER,
    };
}

export const handleOnLogin = (username, password) => (dispatch, getState) => {
    const { users } = getState();

    const user = Object.values(users).find((user) => user.id === username && user.password === password);

    if (!!user) {
        return dispatch(setAuthedUser(user));
    }
}

export const handleOnLogout = () => {
    return (dispatch) => {
        return dispatch(logoutAuthedUser());
    };
}