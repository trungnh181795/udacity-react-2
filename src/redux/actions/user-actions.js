import { ADD_ANSWER_USER, ADD_QUESTION_USER, RECEIVE_USERS } from "../constants/user-constants";

export const receiveUsers = (users) => {
    return {
        type: RECEIVE_USERS,
        users,
    };
}

export const addAnswerUser = (authedUser, qid, answer) => {
    return {
        type: ADD_ANSWER_USER,
        authedUser,
        qid,
        answer,
    };
}

export const addQuestionUser = ({ author, id }) => {
    return {
        type: ADD_QUESTION_USER,
        author,
        qid: id,
    };
}