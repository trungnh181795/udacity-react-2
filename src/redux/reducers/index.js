import { combineReducers } from "redux";
import authReducer from './auth-reducer';
import userReducer from './user-reducer';
import questionsReducer from './question-reducer';

export default combineReducers({
    authedUser: authReducer,
    users: userReducer,
    questions: questionsReducer,
});