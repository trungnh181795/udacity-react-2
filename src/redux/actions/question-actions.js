import { saveQuestion, saveQuestionAnswer } from "../../api/api";
import { ADD_ANSWER_QUESTION, ADD_QUESTION, RECEIVE_QUESTIONS } from "../constants/questions-constants";
import { addAnswerUser, addQuestionUser } from "./user-actions";

export const receiveQuestions = (questions) => {
    return {
        type: RECEIVE_QUESTIONS,
        questions,
    };
}

const addQuestion = (question) => {
    return {
        type: ADD_QUESTION,
        question,
    };
}

const addAnswerQuestion = (author, qid, answer) => {
    return {
        type: ADD_ANSWER_QUESTION,
        author,
        qid,
        answer,
    };
}

export const handleAddQuestion = (firstOption, secondOption) => {
    return async (dispatch, getState) => {
        const { authedUser } = getState();

        return saveQuestion(firstOption, secondOption, authedUser)
            .then((question) => {
                dispatch(addQuestion(question));
                dispatch(addQuestionUser(question))
            })
    };
}

export const handleAddAnswer = (questionId, answer) => {
    return async (dispatch, getState) => {
        const { authedUser } = getState();
        return saveQuestionAnswer(authedUser.id, questionId, answer)
            .then(() => {
                dispatch(addAnswerQuestion(authedUser.id, questionId, answer));
                dispatch(addAnswerUser(authedUser.id, questionId, answer));
            });
    };
}