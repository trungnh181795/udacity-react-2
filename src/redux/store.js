import { legacy_createStore } from "redux";
import reducers from "./reducers";
import middleware from "./middleware";

export const store = legacy_createStore(reducers, middleware)