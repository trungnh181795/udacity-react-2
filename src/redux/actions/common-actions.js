import { getInitialData } from "../../api/api";
import { receiveQuestions } from "./question-actions";
import { receiveUsers } from "./user-actions";

export const handleInitialData = () => {
    return async (dispatch) => {
        return getInitialData().then(({users, questions}) => {
            dispatch(receiveUsers(users));
            dispatch(receiveQuestions(questions));
        });
    };
}