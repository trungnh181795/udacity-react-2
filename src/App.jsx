import '@fontsource/roboto/300.css';
import '@fontsource/roboto/400.css';
import '@fontsource/roboto/500.css';
import '@fontsource/roboto/700.css';
import './App.css';
import { Routes, Route, Navigate } from 'react-router-dom';
import Notfound from './pages/Notfound';
import CreatePoll from './pages/CreatePoll/CreatePoll';
import Poll from './pages/Poll';
import Leaderboard from './pages/Leaderboard';
import Dashboard from './pages/Dashboard';
import Login from './pages/Login';
import { Stack } from '@mui/material';
import { connect } from 'react-redux';
import { useEffect } from 'react';
import { handleInitialData } from './redux/actions/common-actions';
import Navbar from './components/Navbar/Navbar';

function App({ dispatch, loggedIn }) {
  useEffect(() => {
    dispatch(handleInitialData());
  });

  return (
    <>
      {loggedIn ? <Navbar /> : null}
      <Stack justifyContent="center" alignItems="center" sx={{ width: 'calc(100vw - 32px)', height: 'calc(100vh - 70px)', p: 2 }} >
        <Routes>
          <Route path="/login" exact element={<Login />} />
          <Route path="/" element={<Dashboard />} />
          <Route path="/leaderboard" exact element={<Leaderboard />} />
          <Route path="/questions/:id" element={<Poll />} />
          <Route path="/new" exact element={<CreatePoll />} />
          <Route path="/notfound" exact element={<Notfound />} />
          <Route path='*' exact={true} element={<Notfound />} />
        </Routes>
      </Stack>
    </>
  );
}


const mapStateToProps = ({ authedUser }) => ({
  loggedIn: !!authedUser,
});

export default connect(mapStateToProps)(App);
