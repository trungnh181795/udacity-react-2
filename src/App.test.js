/* eslint-disable testing-library/prefer-screen-queries */
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from './redux/store';
import { BrowserRouter as Router } from "react-router-dom";
import App from "./App";
import { setAuthedUser } from "./redux/actions/auth-actions";

describe("App", () => {
  it("should render the view", () => {
    const view = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  it("should show Login page when not logged in", () => {
    const view = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );
    const heading = view.getByTestId("login-header");
    expect(heading).toBeInTheDocument();
  });

  it("should show Dashboard page if user is logged in", () => {
    store.dispatch(setAuthedUser({ id: "sarahedo", password: "password123" }));

    const view = render(
      <Provider store={store}>
        <Router>
          <App />
        </Router>
      </Provider>
    );

    const heading = view.getByTestId("dashboard-header");
    expect(heading).toBeInTheDocument();
  });
});