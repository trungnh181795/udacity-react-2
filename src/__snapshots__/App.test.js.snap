// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the view 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-4angz2-MuiStack-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-h3 css-1cavsrc-MuiTypography-root"
          data-testid="login-header"
        >
          Login
        </h1>
        <form>
          <div
            class="MuiStack-root css-l6w30c-MuiStack-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-14u90gq-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ghx75c-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="username"
              >
                Username
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-d4u71u-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  id="username"
                  name="username"
                  type="text"
                  value="zoshikanlu"
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-1dy8wvs-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ghx75c-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="password"
              >
                Password
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-d4u71u-MuiInputBase-root-MuiInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                  id="password"
                  name="password"
                  type="password"
                  value="pass246"
                />
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-ptgup4-MuiButtonBase-root-MuiButton-root"
              id="login-submit"
              tabindex="0"
              type="submit"
            >
              Login
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-4angz2-MuiStack-root"
    >
      <h1
        class="MuiTypography-root MuiTypography-h3 css-1cavsrc-MuiTypography-root"
        data-testid="login-header"
      >
        Login
      </h1>
      <form>
        <div
          class="MuiStack-root css-l6w30c-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth css-14u90gq-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ghx75c-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="username"
            >
              Username
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-d4u71u-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id="username"
                name="username"
                type="text"
                value="zoshikanlu"
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth css-1dy8wvs-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-1ghx75c-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="password"
            >
              Password
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-d4u71u-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id="password"
                name="password"
                type="password"
                value="pass246"
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-ptgup4-MuiButtonBase-root-MuiButton-root"
            id="login-submit"
            tabindex="0"
            type="submit"
          >
            Login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
